ansible/library/reduce_and_add_sonic_images.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/python
+
+DOCUMENTATION = '''
+module:  reduce_and_add_sonic_images
+version_added:  "1.0"
+
+short_description: remove excessive sonic images install new image if specified
+description: remove excessive sonic images from the target device.
+Note that this version doesn't guarantee to remove older images. Images
+in the 'Available' list that are not 'Current' or 'Next' wil subject to
+removal.
+
+Options:
+    - option-name: disk_used_pcent
+      description: maximum disk used percentage after removing old images
+      required: False
+      Default: 50
+    - option-name: new_image_url
+      description: url pointing to the new image
+      required: False
+      Default: None
+
+'''
+
+import sys
+from ansible.module_utils.basic import *
+
+def exec_command(module, cmd, ignore_error=False, msg="executing command"):
+    rc, out, err = module.run_command(cmd)
+    if not ignore_error and rc != 0:
+        module.fail_json(msg="Failed %s: rc=%d, out=%s, err=%s" %
+                         (msg, rc, out, err))
+    return out
+
+
+def get_sonic_image_list(module):
+    keep   = set()
+    images = set()
+
+    out = exec_command(module, cmd="sonic_installer list",
+                       msg="listing sonic images")
+
+    lines = out.split('\n')
+    for line in lines:
+        line = line.strip()
+        if line.startswith("Current:") or line.startswith("Next:"):
+            keep.add(line.split()[1].strip())
+        elif line != "Available:" and len(line) > 0:
+            images.add(line)
+
+    return (images - keep)
+
+
+def get_disk_free_size(module, partition):
+    out   = exec_command(module, cmd="df -BM --output=avail %s" % partition,
+                         msg="checking disk available size")
+    avail = int(out.split('\n')[1][:-1])
+
+    return avail
+
+
+def get_disk_used_percent(module, partition):
+    out   = exec_command(module, cmd="df -BM --output=pcent %s" % partition,
+                         msg="checking disk available percent")
+    pcent = int(out.split('\n')[1][:-1])
+
+    return pcent
+
+
+def reduce_sonic_image_copies(module, disk_used_pcent):
+    images = get_sonic_image_list(module)
+
+    while len(images) > 0:
+        pcent = get_disk_used_percent(module, "/host")
+        if pcent < disk_used_pcent:
+            break
+        img = images.pop()
```

Comment:
I believe `set.pop()` is non-deterministic because a set is not ordered, so here will we delete a random image.

Wouldn't it make more sense to delete the oldest image?

Reply:
This was intentional.

On systems with only Jenkins built images installed, your suggestion makes sense. However, there are DUTs has developer built images installed, these images doesn't compare too well with the official builds.

With some randomness, Some older images would be removed. 

This is a great question though. I added some comments to clarify the intention.

Reply:
Understood. Maybe in the future we can write a file containing the timestamp of image installation which we can use to determine the oldest image. Just a thought.

Reply:
I think this is a great idea!